### PR TITLE
Fix/4556

### PR DESCRIPTION
**Pull request by zah**
This was quite tricky to get right.

Take a look at the test case in the last commit to see what kinds of new checks are performed.

### CODE REVIEW

File: compiler/semexprs.nim [hunk 1]
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
...
-  case n.kind
+type
+  InitStatus = enum
+    initUnknown
+    initFull     # All  of the fields have been initialized
+    initPartial  # Some of the fields have been initialized
+    initNone     # None of the fields have been initialized
+    initConflict # Fields from different branches have been initialized
+
+proc mergeInitStatus(existing: var InitStatus, newStatus: InitStatus) =
```
Comment (Araq) [inline compiler/semexprs.nim]: This is now complex enough so that it should be put into a ``semobjconstr.nim`` include file.

Comment (zah) [inline compiler/semexprs.nim]: Alright, but let's first merge the fix as it is. I'm already working on something else and juggling these branches takes away from my limited time.

Comment (Araq) [inline compiler/semexprs.nim]: Oh it's not necessary to have a branch per bugfix. Just a single "zahary" branch that you request me to merge from time to time is enough.

File: compiler/semexprs.nim [hunk 2]
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
...
+  # include only a single field
+  if recList.kind == nkSym:
+    yield recList
+  else:
+    internalAssert recList.kind == nkRecList
+    for field in recList:
+      if field.kind != nkSym: continue
+      yield field
+
+template quoeStr(s: string): string = "'" & s & "'"
```
Comment (Araq) [inline compiler/semexprs.nim]: ``quoteStr``

File: compiler/semexprs.nim [hunk 3]
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
...
+
+proc fieldsPresentInInitExpr(fieldsRecList, initExpr: PNode): string =
+  result = ""
+  for field in directFieldsInRecList(fieldsRecList):
+    let assignment = locateFieldInInitExpr(field.sym, initExpr)
+    if assignment != nil:
+      if result.len != 0: result.add ", "
+      result.add field.sym.name.s.quoeStr
```
Comment (Araq) [inline compiler/semexprs.nim]: quoteStr

File: compiler/semexprs.nim [hunk 4]
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
...
-      message(info, errGenerated, "field not initialized: " & n.sym.name.s)
-  else: internalError(info, "checkInitialized")
+    let field = recNode.sym
+    let e = semConstrField(c, flags, field, initExpr)
+    result = if e != nil: initFull else: initNone
+
+  else:
+    internalAssert false
+
+proc semContuctType(c: PContext, t: PType,
```
Comment (Araq) [inline compiler/semexprs.nim]: semConstructType

**Comment by Araq** [comment]
Sorry you need to rebase this. And when you do that, extract code into ``semobjconstr.nim``

**Comment by zah** [comment]
Any objections against pushing it straight to devel after rebasing and
addressing the comments?

**Comment by Araq** [comment]
No. I haven't reviewed it completely but it's localized and I know what it is about. Maybe add a comment about why you take this many passes over the object constructor.

**Comment by vegansk** [comment]
https://github.com/vegansk/nimboost is broken after this

**Pull request merged**